protographic/src/sdl-to-proto-visitor.ts

```diff
@@ -1413,117 +1417,190 @@ Example:
       }
       this.usesWrapperTypes = true; // Track that we're using wrapper types
-      return SCALAR_WRAPPER_TYPE_MAP[graphqlType.name] || 'google.protobuf.StringValue';
+      return {
+        typeName: SCALAR_WRAPPER_TYPE_MAP[graphqlType.name] || 'google.protobuf.StringValue',
+        isRepeated: false,
+      };
     }
 
     if (isEnumType(graphqlType)) {
-      return graphqlType.name;
+      return { typeName: graphqlType.name, isRepeated: false };
     }
 
     if (isNonNullType(graphqlType)) {
       // For non-null scalar types, use the base type
       if (isScalarType(graphqlType.ofType)) {
-        return SCALAR_TYPE_MAP[graphqlType.ofType.name] || 'string';
+        return { typeName: SCALAR_TYPE_MAP[graphqlType.ofType.name] || 'string', isRepeated: false };
       }
 
       return this.getProtoTypeFromGraphQL(graphqlType.ofType);
     }
+    // Named types (object, interface, union, input)
+    const namedType = graphqlType as GraphQLNamedType;
+    if (namedType && typeof namedType.name === 'string') {
+      return { typeName: namedType.name, isRepeated: false };
+    }
 
-    if (isListType(graphqlType)) {
-      // Handle nested list types (e.g., [[Type]])
-      const innerType = graphqlType.ofType;
+    return { typeName: 'string', isRepeated: false }; // Default fallback
+  }
 
-      // If the inner type is also a list, we need to use a wrapper message
-      if (isListType(innerType) || (isNonNullType(innerType) && isListType(innerType.ofType))) {
-        // Find the most inner type by unwrapping all lists and non-nulls
-        let currentType: GraphQLType = innerType;
-        while (isListType(currentType) || isNonNullType(currentType)) {
-          currentType = isListType(currentType) ? currentType.ofType : (currentType as any).ofType;
-        }
+  /**
+   * Converts GraphQL list types to appropriate Protocol Buffer representations.
+   *
+   * For non-nullable, single-level lists (e.g., [String!]!), generates simple repeated fields.
+   * For nullable lists (e.g., [String]) or nested lists (e.g., [[String]]), creates wrapper
+   * messages to properly handle nullability in proto3.
+   *
+   * Examples:
+   * - [String!]! → repeated string field_name = 1;
+   * - [String] → ListOfString field_name = 1; (with wrapper message)
+   * - [[String!]!]! → ListOfListOfString field_name = 1; (with nested wrapper messages)
+   * - [[String]] → ListOfListOfString field_name = 1; (with nested wrapper messages)
+   *
+   * @param type - The GraphQL list type to convert
+   * @returns ProtoType object containing the type name and whether it should be repeated
+   */
+  private handleListType(type: GraphQLList<GraphQLType>): ProtoType {
+    const listType = (isNonNullType(type) ? type.ofType : type) as GraphQLList<GraphQLType>;
+    const isNullableList = !isNonNullType(type);
+    const isNestedList = this.isNestedListType(listType);
```

Comment:
I think this flattening of `GraphQLList<GraphQLType> | GraphQLNonNull<GraphQLList<GraphQLType>>` should happen in a helper function or at the call site, there are assertions both here and at call site that make the exact type of `type` unclear. I would also rename `type` to `graphqlType` like the calling function to avoid shadowing the type keyword